### PR TITLE
Refactoring Laboratories

### PR DESCRIPTION
**Pull request by palexster**
# Description

This PR is the first of a series which refactors the code of the laboratory-operator, making it more compliant with go standard structure.

Test 

- [x] Continous Integration

### CODE REVIEW

**Review by frisso — APPROVED** [review]

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -22,10 +22,19 @@ jobs:
...
           go-version: 1.15
 
+      - name: Install Kubebuilder
+        run: |
+          version=2.3.1 # latest stable version
+          arch=amd64
+          curl -L -O "https://github.com/kubernetes-sigs/kubebuilder/releases/download/v${version}/kubebuilder_${version}_linux_${arch}.tar.gz"
+          tar -zxvf kubebuilder_${version}_linux_${arch}.tar.gz
+          mv kubebuilder_${version}_linux_${arch} kubebuilder && sudo mv kubebuilder /usr/local/
+          export PATH=$PATH:/usr/local/kubebuilder/bin
```
Comment (giorio94) [inline .github/workflows/test.yml:32]: If I am not wrong, this line could be deleted without problems

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -22,10 +22,19 @@ jobs:
...
+  
       - name: Perform the tests
-        working-directory: operators/labInstance-operator/
+        working-directory: operators/
         run: |
-          go test ./pkg/...
+          go test ./...
```
Comment (giorio94) [inline .github/workflows/test.yml:37]: Don't know whether it is better to use this or `make test`... WDYT?

File: operators/pkg/controllers/labinstance_controller.go [hunk 3]
```diff
@@ -22,19 +22,18 @@ import (
...
 
-	"github.com/netgroup-polito/CrownLabs/operators/labInstance-operator/pkg/instanceCreation"
+	"github.com/netgroup-polito/CrownLabs/operators/pkg/instanceCreation"
 
 	"github.com/go-logr/logr"
 	"github.com/google/uuid"
-	instancev1alpha1 "github.com/netgroup-polito/CrownLabs/operators/labInstance-operator/api/v1alpha1"
-	virtv1 "github.com/netgroup-polito/CrownLabs/operators/labInstance-operator/kubeVirt/api/v1"
-	templatev1alpha1 "github.com/netgroup-polito/CrownLabs/operators/labInstance-operator/labTemplate/api/v1alpha1"
+	instancev1alpha1 "github.com/netgroup-polito/CrownLabs/operators/api/v1alpha1"
```
Comment (giorio94) [inline operators/pkg/controllers/labinstance_controller.go:29]:
```suggestion
	crownlabsv1alpha1 "github.com/netgroup-polito/CrownLabs/operators/api/v1alpha1"
```

Just a proposal though, if it easy to be done

File: operators/pkg/instanceCreation/creation.go [hunk 4]
```diff
@@ -3,11 +3,10 @@ package instanceCreation
 import (
 	"context"
 	"encoding/base64"
+	templatev1alpha1 "github.com/netgroup-polito/CrownLabs/operators/api/v1alpha1"
```
Comment (giorio94) [inline operators/pkg/instanceCreation/creation.go:6]:
```suggestion
	crownlabsv1alpha1 "github.com/netgroup-polito/CrownLabs/operators/api/v1alpha1"
```

As before

File: operators/README.md [hunk 5]
```diff
@@ -1,53 +1,100 @@
-# CrownLabs operators
```
Comment (giorio94) [inline operators/README.md:1]: Would it be possible to merge a bit the two README files? Just to make it a bit more general (as it is no longer the lab-operator README file), as well as to make sure all instructions still work (many commands appear out of date).

**Review by giorio94 — COMMENTED** [review]
Thank @palexster for this massive improvement.

In addition to the inline comments, I would propose to:
1. regenerate the CRDs of v1 version (`make manifests` should do the trick)
2. update/delete the `PROJECT` file, as it is out-of-date
3. restore the hack folder, as otherwise `make generate` fails

File: operators/README.md [hunk 6]
```diff
@@ -1,53 +1,91 @@
...
+# Only if HW Virtualization is not available
+kubectl create configmap kubevirt-config -n kubevirt --from-literal debug.useEmulation=true
+# Deploy Kubevirt
+kubectl create -f https://github.com/kubevirt/kubevirt/releases/download/${KUBEVIRT_VERSION}/kubevirt-cr.yaml
+```
+
+#### Deployment
+To deploy the LabOperator in your cluster, you have to do the following steps.
+
+First, set the desired values in `operators/labInstance-operator/k8s-manifest-example.env` .
```
Comment (giorio94) [inline operators/README.md:54]:
```suggestion
First, set the desired values in `operators/deploy/laboratory-operator/k8s-manifest-example.env` .
```

File: operators/README.md [hunk 7]
```diff
@@ -1,53 +1,91 @@
...
+
+Then export the environment variables and generate the manifest from the template using:
+```
+export $(xargs < k8s-manifest-example.env)
```
Comment (giorio94) [inline operators/README.md:58]: I would add `cd operators/deploy/laboratory-operator` at the beginning, just for clarity

File: operators/README.md [hunk 8]
```diff
@@ -1,53 +1,91 @@
...
+After having installed Kubevirt in your testing cluster, you have to deploy the Custom Resource Definitions (CRDs) on the target cluster:
 
-### Controller logic
-The logic of the controller should be put under _controllers/labinstance_controller.go_, in the `Reconcile` method.
-When a LabInstance resource is created, the `Reconcile` method is triggered. 
-It is checked if a LabTemplate resource with the associated name exists, and in this case all the resources necessary to launch the Lab are created,
-in particular the `VirtualMachineInstance` CR managed by KubeVirt. 
+```bash
+make install
+make install-lab-template
```
Comment (giorio94) [inline operators/README.md:82]:
```suggestion

```

This should no longer be needed

File: operators/README.md [hunk 9]
```diff
@@ -1,53 +1,91 @@
...
+```
 
-### Run instructions
-To run the application
+Then, you can launch locally your operator:
 
-- **outside** a cluster: run 'make run'
-- **inside** a cluster: create a Pod with `container.image: crownlabs/laboratory-operator`
+```bash
+make run
```
Comment (giorio94) [inline operators/README.md:88]: There is no longer a `make run` target in the makefile, as it would become meaningless when adding new operators.

**Review by giorio94 — COMMENTED** [review]
Just a couple of minor comments about some leftovers...

Could you also regenerate the CRDs using the latest version of `controller-gen`?
And finally, I would propose to delete `operators/deploy/samples/*`, as it is probably a leftover.

**Review by giorio94 — APPROVED** [review]
/lgtm

**Pull request merged**